Virtual network part 2

ef4 opened the pull request.

ef4 committed:
Implementing mapping in virtual network

ef4 committed:
moved enough things to virtualNetwork urlmapping to pass all host tests

ef4 committed:
more passing tests

ef4 committed:
Merge remote-tracking branch 'origin/main' into virtual-network-part-2

ef4 committed:
fixing more realm-server tests

lukemelia committed:
Get tests passing

jurgenwerk committed:
Remove resolvin/unresolving mapped URLs in loader as this is a concern of virtual network now

jurgenwerk committed:
Not needed anymore

jurgenwerk committed:
Not needed

jurgenwerk committed:
Lint fix

jurgenwerk committed:
Not needed

jurgenwerk committed:
realm server middleware needs to know how to map virtual realm urls

jurgenwerk committed:
Merge branch 'main' into virtual-network-part-2

jurgenwerk committed:
URL mapping has moved to virtual network

jurgenwerk committed:
Not needed

jurgenwerk commented:
@ef4 @lukemelia  an update:

1. Cleaned up the concepts of resolving, unresolving, url mapping from the loader (basically what https://github.com/cardstack/boxel/pull/1115 is trying to achieve - it seemed more convenient to do it here in this branch)

2. While hosts and realm tests were passing, the realm dom tests were failing (this could easily be reprouced by going to http://localhost:4201 which was broken because the browser was trying to load assets from an unresolved url, i.e. cardstack.com/base). It was happening because the realm server middleware is doing redirections (`rootRealmRedirect`, `assetRedirect`)  and in that logic it wants to know the resolved URL for the realm. For that reason I changed virtual network's `resolveURLMapping` from `private` to public, and used it in redirection logic: https://github.com/cardstack/boxel/pull/1126/commits/6f74533ac6ccc62406ffdfd5a10573f309989cdc

I feel that this change might not be aligned with the vision for the virtual network so I'd like to discuss this. Perhaps the realm server should not know about resolved urls and the mentioned redirection logic should be moved to the virtual network as handlers?

3. At this point there are still test failures, this time in [matrix tests](https://github.com/cardstack/boxel/actions/runs/8523165799/job/23344878244?pr=1126). There seems to be an issue with fetching rooms from the matrix server where it doesn't return the chat room. It's puzzling how this is related to this refactor but I am checking it.

jurgenwerk commented:
Follow up comments to issue described in (2), did some discussing with @lukemelia–a couple of ideas to make the `resolveURLMapping` back to being a private method:

1. There could be an `AssetRedirector` that would receive the mapping in the constructor, and have a method to be plugged in the middleware (similarly to `assetRedirect`). It would respond with the real url in the location header by figuring it out from the url mapping config rather than calling `resolveURLMapping`  

2. We could explore mounting a handler for the redirection logic (as opposed to doing that in the koa middleware), mount it to virtual network, and add functionality in `handle` which would add special treatment if the response results in a redirect (making the location header url real, not virtual)

jurgenwerk committed:
Address a missing duplex property when remapping a request

jurgenwerk committed:
Do not make a new request if the mapped url has not changed

This fixes an error when closeHandlers.get(request)!.on('close', fn) errors out  because the copied request can't be found in the closeHandlers weak map

jurgenwerk committed:
Setup close handler for the mapped request

jurgenwerk committed:
Deal with issues of mapping a request that has a body

jurgenwerk committed:
Merge branch 'main' into virtual-network-part-2

jurgenwerk committed:
When a handler in virtual network returns a redirect make sure to give a real url (not virtual)

jurgenwerk committed:
Move url mapping responsibility from realm server to virtual network

jurgenwerk committed:
Merge branch 'main' into virtual-network-part-2

jurgenwerk committed:
Make resolveURLMapping in virtual network private again

jurgenwerk commented:
This PR leverages the virtual network in such a way that:

1. All URL mapping is moved to the virtual network which is now fully responsible for mapping (also known as resolving, reverse resolving realm urls) 
2. All logic that depended on knowing the resolved url was moved to the virtual network 
3. Redirections (asset redirection, realm root redirection) were moved out of the realm server (to satisfy (1)) 

This refactor is an intermediate step towards having isolated loaders that share a virtual network.

lukemelia commented on packages/host/app/services/message-service.ts:
```diff
@@ -1,7 +1,8 @@
 import { tracked } from '@glimmer/tracking';
 
-import type LoaderService from './loader-service';
+import LoaderService from './loader-service';
```
Seems like this should still be a type import

jurgenwerk replied:
Fixed.

lukemelia commented on packages/realm-server/main.ts:
```diff
@@ -175,33 +175,33 @@ if (distURL) {
   }
 
-  let server = new RealmServer(realms, {
+  let server = new RealmServer(realms, virtualNetwork, {
```
I may be reading it wrong, but it seems like the realm server is already mounting it's maybeExternalHandle method, so why do we also mount it on line 201?

jurgenwerk replied:
Hmm, I see that on line 201 it's the only time the realm server is mounting `maybeExternalHandle`. The realm server is mounting `maybeHandle` - did you mean that?

lukemelia replied:
I must have been confused.

lukemelia commented on packages/runtime-common/realm.ts:
```diff
@@ -968,30 +1000,24 @@ export class Realm {
-        ],
-      ]);
-    
+    let endpontsWithoutAuthNeeded: RouteTable<true> = new Map([
```
typo: should be `endpointsWithoutAuthNeeded`

jurgenwerk replied:
Done 👍

lukemelia commented on packages/runtime-common/virtual-network.ts:
```diff
@@ -67,6 +115,73 @@ export class VirtualNetwork {
+    return response;
+  };
+
+  async handle(
```
Add a comment explaining what this method should be used for.

jurgenwerk replied:
Added a comment.

lukemelia replied:
I don't see the comment.

lukemelia replied:
nvm, I see it now!

lukemelia commented on packages/runtime-common/virtual-network.ts:
```diff
@@ -86,3 +206,63 @@ function isUrlLike(moduleIdentifier: string): boolean {
+
+  let body = null;
+  if (originalRequest.body) {
+    body = await getContentOfReadableStream(originalRequest.clone().body);
```
How does this work with SSE, or does it need to? My understanding of SSE is that the body never is "finished".

habdelra replied:
Native browser API is used to subscribe to an SSE event--we don't go thru a fetch (https://github.com/cardstack/boxel/blob/main/packages/host/app/services/message-service.ts#L19). Given that I wonder if you still need to unresolve that URL that you pass to the `new EventSource(url)` that we use to subscribe to SSE events. Right now only the base realm really has a URL that is different between the resolved and unresolved forms--so if we are not subscribing to the base realm SSE (and I don't think we ever do), it's not something that you'd be able to see (so like a hidden bug).

jurgenwerk replied:
That's right, SSE uses an `EventSource` - https://github.com/cardstack/boxel/pull/1126/files#diff-cfc037186d9054c397cd997d5445153953255681ca1d5e29b02d4278345763b3R20

Regarding subscribing to base realm using SSE - I observed that we actually do subscribe to the base realm SSE, for example if you visit `https://cardstack.com/base/boolean.gts` then the file resource will subscribe to the base realm.

habdelra replied:
If you change the Boolean.gts file in the base realm does the app fetch the new change? I think it is literally subscribed to cardstack.com not the base realm running on localhost.

jurgenwerk replied:
Yes, the change is reflected in the host code editor if I change the boolean file locally. It will correctly connect to the resolved url via event source. So I think this is all good?

habdelra reviewed (approved):
nice!

jurgenwerk committed:
Can narrow down to type import

jurgenwerk committed:
Typo

jurgenwerk committed:
Add a comment about the usage of handle method in virtualNetwork

jurgenwerk committed:
Not needed here anymore

lukemelia committed:
Extract PackageShimHandler as its own simple class instead of using Loader

jurgenwerk committed:
Merge pull request #1162 from cardstack/package-shim-handler

Extract PackageShimHandler as its own simple class instead of using Loader

jurgenwerk committed:
Merge branch 'main' into virtual-network-part-2

jurgenwerk commented on packages/runtime-common/virtual-network.ts:
```diff
@@ -86,3 +205,63 @@ function isUrlLike(moduleIdentifier: string): boolean {
+
+  let body = null;
+  if (originalRequest.body) {
+    body = await getContentOfReadableStream(originalRequest.clone().body);
```
I've been trying to see if I can get rid of body stream reading to create a new request with a changed url but to no avail...

1. `Object.defineProperty(request, 'url', { value: mappedUrl })` will not work because the underlying url value really is immutable. Here is a demonstration - the request will still try to reach the original url even when trying to override it with `defineProperty`: 

<img width="579" alt="image" src="https://github.com/cardstack/boxel/assets/273660/90d5cd86-852b-4c28-829b-bd214e8edc8d">

2. I saw some hints online that [tee](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/tee) could be used to copy the body, like so:

```
let copiedBody = null;
  if (originalRequest.body) {
    copiedBody = originalRequest.body.tee()[0];
  }

  return new Request(url, {
    method: originalRequest.method,
    headers: originalRequest.headers,
    body: copiedBody,
    referrer: originalRequest.referrer,
    referrerPolicy: originalRequest.referrerPolicy,
    mode: originalRequest.mode,
    credentials: originalRequest.credentials,
    cache: originalRequest.cache,
    redirect: originalRequest.redirect,
    integrity: originalRequest.integrity,
  });
  ```
  
  
  But again, the error persists:
  
<img width="1170" alt="image" src="https://github.com/cardstack/boxel/assets/273660/88c3efb1-021f-49da-9c7f-16ae69663964">

3. `clone()`-ing the request and taking its body won't work either (failing with the same error): 

```
return new Request(url, {
    method: originalRequest.method,
    headers: originalRequest.headers,
    body: originalRequest.clone().body,
    referrer: originalRequest.referrer,
    referrerPolicy: originalRequest.referrerPolicy,
    mode: originalRequest.mode,
    credentials: originalRequest.credentials,
    cache: originalRequest.cache,
    redirect: originalRequest.redirect,
    integrity: originalRequest.integrity,
  });
```

It is interesting because Safari gives a different error message:

<img width="415" alt="image" src="https://github.com/cardstack/boxel/assets/273660/e1ce8d02-e6f2-4a26-bff8-6ed938541391">


4. But what if we try adding the `duplex` property? 

```
let requestInit: RequestInit = originalRequest.clone();

  if (originalRequest.body) {
    requestInit.duplex = 'half';
  }

  return new Request(url, requestInit);
```

The request will fail with `net::ERR_H2_OR_QUIC_REQUIRED` which is indicative of missing http2 support on the server:

<img width="605" alt="image" src="https://github.com/cardstack/boxel/assets/273660/6da127fa-62fa-4b4b-879d-2002a790fa79">

Moreover, TypeScript doesn't recognize `duplex` property on the `Request` - looks like this is more of an experimental thing for now...

habdelra replied:
Another thought would be to Proxy the request: 
```ts 
let proxiedRequest = new Proxy(originalRequest, {
  get(target, property, received) {
    if (property === 'url') {
      return mappedURL;
    }
    return Reflect.get(target, property, received);
  }
})

jurgenwerk committed:
Fix typo
